### Spark Moive Recommendation
In this notebook, we will use an Alternating Least Squares (ALS) algorithm with Spark APIs to predict the ratings for the movies in [MovieLens small dataset](https://grouplens.org/datasets/movielens/latest/)

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import math
%matplotlib inline

%matplotlib inline is not supported in Databricks.
You can display matplotlib figures using display(). For an example, see https://docs.databricks.com/user-guide/visualizations/matplotlib-and-ggplot.html

In [3]:
import os
os.environ["PYSPARK_PYTHON"] = "python3"

## Part1: Data ETL and Data Exploration

In [5]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("moive analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [6]:
movies_df = spark.read.load("/FileStore/tables/movies.csv", format='csv', header = True)
ratings_df = spark.read.load("/FileStore/tables/ratings.csv", format='csv', header = True)
links_df = spark.read.load("/FileStore/tables/links.csv", format='csv', header = True)
tags_df = spark.read.load("/FileStore/tables/tags.csv", format='csv', header = True)

In [7]:
movies_df.show(5)

+-------+--------------------+--------------------+
movieId| title| genres|
+-------+--------------------+--------------------+
 1| Toy Story (1995)|Adventure|Animati...|
 2| Jumanji (1995)|Adventure|Childre...|
 3|Grumpier Old Men ...| Comedy|Romance|
 4|Waiting to Exhale...|Comedy|Drama|Romance|
 5|Father of the Bri...| Comedy|
+-------+--------------------+--------------------+
only showing top 5 rows

In [8]:
ratings_df.show(5)

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
+------+-------+------+---------+
only showing top 5 rows

In [9]:
links_df.show(5)

+-------+-------+------+
movieId| imdbId|tmdbId|
+-------+-------+------+
 1|0114709| 862|
 2|0113497| 8844|
 3|0113228| 15602|
 4|0114885| 31357|
 5|0113041| 11862|
+-------+-------+------+
only showing top 5 rows

In [10]:
tags_df.show(5)

+------+-------+---------------+----------+
userId|movieId| tag| timestamp|
+------+-------+---------------+----------+
 2| 60756| funny|1445714994|
 2| 60756|Highly quotable|1445714996|
 2| 60756| will ferrell|1445714992|
 2| 89774| Boxing story|1445715207|
 2| 89774| MMA|1445715200|
+------+-------+---------------+----------+
only showing top 5 rows

## Part I: Exploratory Data Analysis

In [12]:
tmp1 = ratings_df.groupBy("userID").count().toPandas()['count'].min()
tmp2 = ratings_df.groupBy("movieId").count().toPandas()['count'].min()
print('For the users that rated movies and the movies that were rated:')
print('Minimum number of ratings per user is {}'.format(tmp1))
print('Minimum number of ratings per movie is {}'.format(tmp2))

/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "
/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "
For the users that rated movies and the movies that were rated:
Minimum number of ratings per user is 20
Minimum number of ratings per movie is 1

In [13]:
tmp1 = sum(ratings_df.groupBy("movieId").count().toPandas()['count'] == 1)
tmp2 = ratings_df.select('movieId').distinct().count()
print('{} out of {} movies are rated by only one user'.format(tmp1, tmp2))

/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "
3446 out of 9724 movies are rated by only one user

In [14]:
tmp1 = ratings_df.groupBy("movieId").count().toPandas()['count'].mean()
tmp2 = ratings_df.groupBy("userId").count().toPandas()['count'].mean()
print('Average number of ratings per movie is {}'.format(tmp1))
print('Average number of ratings per user is {}'.format(tmp2))

/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "
/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "
Average number of ratings per movie is 10.369806663924312
Average number of ratings per user is 165.30491803278687

In [15]:
movies_df.registerTempTable("movies")
ratings_df.registerTempTable("ratings")
links_df.registerTempTable("links")
tags_df.registerTempTable("tags")

### The number of Users

In [17]:
%sql 
SELECT Count(DISTINCT userId) AS number_of_users 
FROM ratings

number_of_users 610

###The number of Movies

In [19]:
%sql 
SELECT Count(DISTINCT movieID) AS number_of_movies
FROM movies

number_of_movies 9742

### The sparsity of the movie ratings

### sparcity = 1- rating_num/(movie_num*user_num)

In [21]:
movie_num = ratings_df.select('movieId').distinct().count()
user_num = ratings_df.select('userId').distinct().count()
rating_num = ratings_df.select('rating').count()

In [22]:
denominator = movie_num*user_num
numerator = rating_num
sparsity = (1-numerator/denominator)*100
print ("The ratings dataframe is ", "%.2f" % sparsity + "% empty.")

The ratings dataframe is 98.30% empty.

### Movies rated by users.

In [24]:
%sql 
SELECT Count(DISTINCT movieID) AS movies_rated
FROM ratings 

movies_rated 9724

### Movies not rated before.

In [26]:
%sql
SELECT m.*,r.rating
FROM movies m 
    LEFT JOIN ratings r ON m.movieId = r.movieId
WHERE r.rating is Null

movieId title genres rating 1076 Innocents, The (1961) Drama|Horror|Thriller null 2939 Niagara (1953) Drama|Thriller null 3338 For All Mankind (1989) Documentary null 3456 Color of Paradise, The (Rang-e khoda) (1999) Drama null 4194 I Know Where I'm Going! (1945) Drama|Romance|War null 5721 Chosen, The (1981) Drama null 6668 Road Home, The (Wo de fu qin mu qin) (1999) Drama|Romance null 6849 Scrooge (1970) Drama|Fantasy|Musical null 7020 Proof (1991) Comedy|Drama|Romance null 7792 Parallax View, The (1974) Thriller null 8765 This Gun for Hire (1942) Crime|Film-Noir|Thriller null 25855 Roaring Twenties, The (1939) Crime|Drama|Thriller null 26085 Mutiny on the Bounty (1962) Adventure|Drama|Romance null 30892 In the Realms of the Unreal (2004) Animation|Documentary null 32160 Twentieth Century (1934) Comedy null 32371 Call Northside 777 (1948) Crime|Drama|Film-Noir null 34482 Browning Version, The (1951) Drama null 85565 Chalet Girl (2011) Comedy|Romance null

### List Movie Genres

Each movie belongs to more than 1 genre,as shown below. We need to seperate these genres.

In [29]:
%sql
SELECT DISTINCT title, genres
FROM movies

title genres Color of Night (1994) Drama|Thriller Surviving the Game (1994) Action|Adventure|Thriller Little Big League (1994) Comedy|Drama Whatever (1998) Drama Ronin (1998) Action|Crime|Thriller Pink Flamingos (1972) Comedy Austin Powers: The Spy Who Shagged Me (1999) Action|Adventure|Comedy Erin Brockovich (2000) Drama Porky's Revenge (1985) Comedy Angel Heart (1987) Film-Noir|Horror|Mystery|Thriller D.O.A. (1988) Film-Noir|Mystery|Thriller Extreme Days (2001) Action|Adventure|Comedy|Drama MacArthur (1977) Drama Madame Bovary (1991) Drama Body of Evidence (1993) Drama|Thriller War and Peace (1956) Drama|Romance|War Flight of the Phoenix, The (1965) Action|Adventure|Drama My Boss's Daughter (2003) Comedy|Romance Wonderland (2003) Crime|Drama|Mystery|Thriller Intolerance: Love's Struggle Throughout the Ages (1916) Drama Virgin Spring, The (Jungfrukällan) (1960) Crime|Drama Shaun of the Dead (2004) Comedy|Horror Primer (2004) Drama|Sci-Fi Snow Walker, The (2003) Adventure|Drama Short Film About Love, A (Krótki film o milosci) (1988) Drama|Romance Matador, The (2005) Comedy|Drama|Thriller District 13 (Banlieue 13) (2004) Action|Crime|Sci-Fi Let's Go to Prison (2006) Comedy Lucky You (2007) Comedy|Drama Awake (2007) Drama|Thriller Legionnaire (1998) Action|Adventure|Drama|War Up in the Air (2009) Drama|Romance Date Night (2010) Action|Comedy|Romance Iron Man 2 (2010) Action|Adventure|Sci-Fi|Thriller|IMAX Winter's Bone (2010) Drama|Thriller Knight and Day (2010) Action|Comedy|Romance Last Exorcism, The (2010) Horror|Thriller Going the Distance (2010) Comedy|Romance What's Your Number? (2011) Comedy|Romance Safe Haven (2013) Drama|Mystery|Romance Epic (2013) Adventure|Animation|Fantasy Justice League: Crisis on Two Earths (2010) Action|Animation|Sci-Fi Black Sea (2015) Adventure|Mystery|Thriller Terminator Genisys (2015) Action|Adventure|Sci-Fi|Thriller The Duke of Burgundy (2014) Drama Rock the Kasbah (2015) Comedy Winter in Prostokvashino (1984) Animation Young Poisoner's Handbook, The (1995) Crime|Drama Awfully Big Adventure, An (1995) Drama Blink (1994) Thriller Coneheads (1993) Comedy|Sci-Fi So I Married an Axe Murderer (1993) Comedy|Romance|Thriller Hype! (1996) Documentary Head Above Water (1996) Comedy|Thriller Sliding Doors (1998) Drama|Romance Madeline (1998) Children|Comedy Dark Crystal, The (1982) Adventure|Fantasy Beetlejuice (1988) Comedy|Fantasy Peeping Tom (1960) Drama|Horror|Thriller Christine (1983) Horror Airport '77 (1977) Drama Killer's Kiss (1955) Crime|Film-Noir Angela's Ashes (1999) Drama Gold Rush, The (1925) Adventure|Comedy|Romance For the Boys (1991) Comedy|Drama|Musical Baby Boy (2001) Crime|Drama Let It Ride (1989) Comedy Polyester (1981) Comedy Chasing Papi (a.k.a. Papi Chulo) (2003) Comedy Once Bitten (1985) Comedy|Horror Beethoven's 2nd (1993) Children|Comedy Undercurrent (1946) Film-Noir|Thriller Sniper (1993) Action|Drama Man Who Came to Dinner, The (1942) Comedy Baadasssss! (How to Get the Man's Foot Outta Your Ass) (2003) Drama Twilight Samurai, The (Tasogare Seibei) (2002) Drama|Romance Dr. Giggles (1992) Comedy|Horror C.R.A.Z.Y. (2005) Drama Stardust (2007) Adventure|Comedy|Fantasy|Romance 30 Days of Night (2007) Horror|Thriller Before the Devil Knows You're Dead (2007) Crime|Drama|Thriller Like Stars on Earth (Taare Zameen Par) (2007) Drama Forgetting Sarah Marshall (2008) Comedy|Romance Winnie the Pooh and Tigger Too (1974) Animation|Children Trance (2013) Crime|Thriller Scouts Guide to the Zombie Apocalypse (2015) Action|Comedy|Horror Self-criticism of a Bourgeois Dog (2017) Comedy What's Eating Gilbert Grape (1993) Drama Barbarella (1968) Adventure|Comedy|Sci-Fi Fled (1996) Action|Adventure It's a Wonderful Life (1946) Children|Drama|Fantasy|Romance Old Man and the Sea, The (1958) Adventure|Drama Apocalypse Now (1979) Action|Drama|War High Noon (1952) Drama|Western Red Corner (1997) Crime|Thriller Twilight (1998) Crime|Drama|Thriller Nightmare on Elm Street 3: Dream Warriors, A (1987) Hor

In [30]:
## Data processing to seperate the genres for a movie
genres_pd_df = spark.sql("SELECT DISTINCT title, genres FROM movies").toPandas()
genres_pd_df['genres'] = genres_pd_df['genres'].apply(lambda x:x.split('|'))
genres_pd_df = pd.concat([genres_pd_df['title'],genres_pd_df['genres'].apply(pd.Series)],axis = 1).set_index('title')
genres_sep_pd_df = genres_pd_df.stack().reset_index(level=0)
genres_sep_pd_df.columns = ['title','genre']

/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "

We store all the movie genres in a list called movie_genres_list.

In [32]:
movie_genres_list = genres_sep_pd_df['genre'].unique().tolist()
print('Here are all the movie genres:'+'\n', movie_genres_list)

Here are all the movie genres:
 ['Drama', 'Thriller', 'Action', 'Adventure', 'Comedy', 'Crime', 'Film-Noir', 'Horror', 'Mystery', 'Romance', 'War', 'Sci-Fi', 'IMAX', 'Animation', 'Fantasy', 'Documentary', 'Children', 'Musical', 'Western', '(no genres listed)']

### Movie for Each Category

In [34]:
genres_sep_df = sqlContext.createDataFrame(genres_sep_pd_df)
genres_sep_df.registerTempTable('genres_sep')

In [35]:
%sql
SELECT genre,count(*) AS count
FROM genres_sep
GROUP BY 1
ORDER BY 2 DESC

genre count Drama 4361 Comedy 3756 Thriller 1894 Action 1828 Romance 1596 Adventure 1263 Crime 1199 Sci-Fi 980 Horror 978 Fantasy 779 Children 664 Animation 611 Mystery 573 Documentary 440 War 382 Musical 334 Western 167 IMAX 158 Film-Noir 87 (no genres listed) 34

We put each movie under its genre. This is stored in a dictionary called movie_genre_dict, where the key is the genre, and the values is a list that contains the movies belonging to this genre.

In [37]:
movie_genre_dict = dict()
for _ in movie_genres_list:
  movie_genre_dict[_] = list()
for index, row in genres_sep_pd_df.iterrows():
   movie_genre_dict[row["genre"]].append(row["title"])

Let's take a look at the crime genre as an example.

In [39]:
print('Here are movies that belong to the crime genre:'+'\n',movie_genre_dict['Crime'])

Here are movies that belong to the crime genre:
 ['Ronin (1998)', 'Wonderland (2003)', 'Virgin Spring, The (Jungfrukällan) (1960)', 'District 13 (Banlieue 13) (2004)', "Young Poisoner's Handbook, The (1995)", "Killer's Kiss (1955)", 'Baby Boy (2001)', "Before the Devil Knows You're Dead (2007)", 'Trance (2013)', 'Red Corner (1997)', 'Twilight (1998)', 'Shanghai Surprise (1986)', 'Badlands (1973)', 'Dobermann (1997)', 'Polytechnique (2009)', 'Mud (2012)', 'Reservoir Dogs (1992)', 'One Tough Cop (1998)', '11:14 (2003)', 'Infernal Affairs 2 (Mou gaan dou II) (2003)', 'Gentlemen of Fortune (Dzhentlmeny udachi) (1972)', 'Rumble in the Bronx (Hont faan kui) (1995)', 'Postman Always Rings Twice, The (1981)', 'John Q (2002)', 'Welcome to Collinwood (2002)', 'Remo Williams: The Adventure Begins (1985)', 'Elite Squad (Tropa de Elite) (2007)', "Let's Be Cops (2014)", 'Once Upon a Time in America (1984)', 'Jackie Brown (1997)', 'Drugstore Cowboy (1989)', 'Snatch (2000)', "Man Who Wasn't There, The (2001)", 'Monster (2003)', 'Killers, The (1946)', "Flatfoot on the Nile (Piedone d'Egitto) (1980)", 'Blitz (2011)', 'Sorrow (2015)', 'Sympathy for the Underdog (1971)', 'Dick Tracy (1990)', 'Dirty Harry (1971)', 'Bourne Supremacy, The (2004)', 'Seven-Per-Cent Solution, The (1976)', 'Mother (Madeo) (2009)', 'Rent-A-Cop (1988)', 'Burglar (1987)', "What's Up, Tiger Lily? (1966)", 'Freaks (1932)', 'Munich (2005)', 'Crime Busters (1977)', 'Getaway (2013)', 'All About the Benjamins (2002)', 'I, the Jury (1982)', 'United States of Leland, The (2003)', 'Dead or Alive: Hanzaisha (1999)', 'Papillon (1973)', 'Thief of Paris, The (Le voleur) (1967)', 'Son of a Gun (2014)', 'Menace II Society (1993)', 'Juice (1992)', 'Performance (1970)', 'Casino Jack (2010)', 'Léon: The Professional (a.k.a. The Professional) (Léon) (1994)', 'Godfather: Part II, The (1974)', 'Black Dahlia, The (2006)', 'Boondock Saints II: All Saints Day, The (2009)', '30 Minutes or Less (2011)', 'Batman: The Killing Joke (2016)', 'Strange Days (1995)', 'Transporter, The (2002)', 'Slumdog Millionaire (2008)', 'Horrible Bosses (2011)', 'In Time (2011)', 'Set It Off (1996)', 'Fried Green Tomatoes (1991)', 'Lost Highway (1997)', 'Belly (1998)', 'Coffy (1973)', 'Blood Work (2002)', 'People I Know (2002)', 'Bugsy (1991)', 'Fantastic Mr. Fox (2009)', 'Return of the Pink Panther, The (1975)', 'Point Blank (1967)', 'School for Scoundrels (2006)', 'Brothers Bloom, The (2008)', 'Oxygen (1999)', 'Cradle 2 the Grave (2003)', 'United 93 (2006)', 'Micmacs (Micmacs à tire-larigot) (2009)', 'Killer Joe (2011)', 'Best Men (1997)', 'Ghost Dog: The Way of the Samurai (1999)', 'Lights in the Dusk (Laitakaupungin valot) (2006)', 'Undisputed II: Last Man Standing (2006)', 'Machete (2010)', 'The Adventures of Sherlock Holmes and Doctor Watson: The Treasures of Agra (1983)', 'Trainspotting (1996)', 'Sexy Beast (2000)', 'City by the Sea (2002)', 'Empire (2002)', 'Better Luck Tomorrow (2002)', 'Assault on Precinct 13 (2005)', 'Tie Me Up! Tie Me Down! (¡Átame!) (1990)', "We're No Angels (1989)", 'Lock, Stock & Two Smoking Barrels (1998)', 'Death Wish (1974)', 'Bob le Flambeur (1955)', 'This Gun for Hire (1942)', 'Cellular (2004)', 'This Must Be the Place (2011)', 'Prime Suspect 3 (1993)', 'Captain America II: Death Too Soon (1979)', 'Go (1999)', 'Hard-Boiled (Lat sau san taam) (1992)', 'Crimson Rivers, The (Rivières pourpres, Les) (2000)', 'V. I. Warshawski (1991)', "Malibu's Most Wanted (2003)", 'Scarface (1932)', 'Perfect Crime, The (Crimen Ferpecto) (Ferpect Crime) (2004)', 'Contraband (2012)', 'True Crime (1999)', 'Shaft (1971)', 'Buffalo Soldiers (2001)', 'Dark Passage (1947)', 'Police Story (Ging chaat goo si) (1985)', 'Stander (2003)', 'Transporter 2 (2005)', 'Bourne Ultimatum, The (2007)', 'Assassination of Jesse James by the Coward Robert Ford, The (2007)', 'Delhi Belly (2011)', 'Alex Cross (2012)', 'Purge, The (2013)', 'Equalizer, The (2014)', 'Nightcrawler (2014)', 'BlacKkKlansman (2018)', 'Crossing Guard, 

## Part2: Spark ALS based approach for training model
We will use an Spark ML to predict the ratings, so let's reload "ratings.csv" using ``sc.textFile`` and then convert it to the form of (user, item, rating) tuples.

In [41]:
ratings_df.show()

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
 1| 70| 3.0|964982400|
 1| 101| 5.0|964980868|
 1| 110| 4.0|964982176|
 1| 151| 5.0|964984041|
 1| 157| 5.0|964984100|
 1| 163| 5.0|964983650|
 1| 216| 5.0|964981208|
 1| 223| 3.0|964980985|
 1| 231| 5.0|964981179|
 1| 235| 4.0|964980908|
 1| 260| 5.0|964981680|
 1| 296| 3.0|964982967|
 1| 316| 3.0|964982310|
 1| 333| 5.0|964981179|
 1| 349| 4.0|964982563|
+------+-------+------+---------+
only showing top 20 rows

In [42]:
movie_ratings_df=ratings_df.drop('timestamp')
movie_ratings_df.show()

+------+-------+------+
userId|movieId|rating|
+------+-------+------+
 1| 1| 4.0|
 1| 3| 4.0|
 1| 6| 4.0|
 1| 47| 5.0|
 1| 50| 5.0|
 1| 70| 3.0|
 1| 101| 5.0|
 1| 110| 4.0|
 1| 151| 5.0|
 1| 157| 5.0|
 1| 163| 5.0|
 1| 216| 5.0|
 1| 223| 3.0|
 1| 231| 5.0|
 1| 235| 4.0|
 1| 260| 5.0|
 1| 296| 3.0|
 1| 316| 3.0|
 1| 333| 5.0|
 1| 349| 4.0|
+------+-------+------+
only showing top 20 rows

In [43]:
# Data type convert
from pyspark.sql.types import IntegerType, FloatType
movie_ratings_df = movie_ratings_df.withColumn("userId", movie_ratings_df["userId"].cast(IntegerType()))
movie_ratings_df = movie_ratings_df.withColumn("movieId", movie_ratings_df["movieId"].cast(IntegerType()))
movie_ratings_df = movie_ratings_df.withColumn("rating", movie_ratings_df["rating"].cast(FloatType()))

In [44]:
movie_ratings_df.show()

+------+-------+------+
userId|movieId|rating|
+------+-------+------+
 1| 1| 4.0|
 1| 3| 4.0|
 1| 6| 4.0|
 1| 47| 5.0|
 1| 50| 5.0|
 1| 70| 3.0|
 1| 101| 5.0|
 1| 110| 4.0|
 1| 151| 5.0|
 1| 157| 5.0|
 1| 163| 5.0|
 1| 216| 5.0|
 1| 223| 3.0|
 1| 231| 5.0|
 1| 235| 4.0|
 1| 260| 5.0|
 1| 296| 3.0|
 1| 316| 3.0|
 1| 333| 5.0|
 1| 349| 4.0|
+------+-------+------+
only showing top 20 rows

### ALS Model Selection and Evaluation

With the ALS model, we can use a grid search to find the optimal hyperparameters.

In [46]:
# import package
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import CrossValidator,ParamGridBuilder

In [47]:
#Create test and train set
(training,test)=movie_ratings_df.randomSplit([0.8,0.2],seed = 42)

## We will tune the hyperparameters using ParamGridBuilder and CrossValidator.

In [49]:
#Create ALS model
als = ALS(userCol="userId", itemCol="movieId", ratingCol="rating",coldStartStrategy = 'drop',nonnegative = True, implicitPrefs = False)
# Confirm that a model called "als" was created
type(als)

Out[26]: pyspark.ml.recommendation.ALS

In [50]:
#Tune model using ParamGridBuilder
# We will just tune rank and regParam considering long run time, after we get the have combination, we will use larger iterations.
param_grid = ParamGridBuilder() \
            .addGrid(als.rank, [3,5,10]) \
            .addGrid(als.maxIter, [10]) \
            .addGrid(als.regParam, [0.05,0.15,0.25]) \
            .build()
print ("Num models to be tested: ", len(param_grid))

Num models to be tested: 9

In [51]:
# Define evaluator as RMSE
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction") 

In [52]:
# Build Cross validation 
cv = CrossValidator(estimator=als, estimatorParamMaps=param_grid, evaluator=evaluator, numFolds=3)

In [53]:
#Fit ALS model to training data
model = cv.fit(training)

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [54]:
#Extract best model from the tuning exercise using ParamGridBuilder
best_model = model.bestModel

In [55]:
#Generate predictions and evaluate using RMSE
predictions=best_model.transform(test)
rmse = evaluator.evaluate(predictions)

In [56]:
#Print evaluation metrics and model parameters
print ("RMSE = "+str(rmse))
print ("**Best Model**")
print (" Rank:",best_model._java_obj.parent().getRank())   #parent()method will return an estimator,you can get the best params then
print (" MaxIter:",best_model._java_obj.parent().getMaxIter())
print (" RegParam:",best_model._java_obj.parent().getRegParam()) 

RMSE = 0.8736197761693623
**Best Model**
 Rank: 10
 MaxIter: 10
 RegParam: 0.15

### Increase iteration number and Model testing

In [58]:
#Increase the interation for the best ALS model
#coldStartStrategy = 'drop' is important, otherwise, you will recieve rmse = nan
#Spark allows users to set the coldStartStrategy parameter to “drop” in order to drop any rows in the DataFrame of predictions that contain NaN values. 
als_50 = ALS(userCol="userId", itemCol="movieId", ratingCol="rating",rank = 10, maxIter = 50,regParam = 0.15, nonnegative = True, coldStartStrategy = 'drop',implicitPrefs = False)
#fit the model to training data
best_model_50 = als_50.fit(training)

In [59]:
#generate predictions on test data
prediction_50 = best_model_50.transform(test)

In [60]:
#tell spark how to evaluate predictions
evaluator_50 = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
#obtain rmse
rmse_50 = evaluator_50.evaluate(prediction_50)
#print rmse
print('RMSE=',rmse_50)

RMSE= 0.8705602687772467

In [61]:
prediction_50_pd_df = prediction_50.toPandas()[['movieId','rating','prediction']].set_index('movieId')
display(prediction_50_pd_df.plot(style=['o','rx']))

In [62]:
# round the prediction to a scale of 1-5
def round_to_5scale(x):
  if x<=round(x) and 0.5<=x<=5:
    return round(x)-0.5 if round(x)-x>0.25 else round(x)
  if x>round(x) and 0.5<=x<=5:
    return round(x)+0.5 if x-round(x)>0.25 else round(x)
  if x>5:
    return 5
  if x<0.5:
    return 0.5

In [63]:
prediction_50_pd_df['prediction'] = prediction_50_pd_df['prediction'].apply(lambda x: round_to_5scale(x))
display(prediction_50_pd_df.plot(style=['o','rx']))

In [64]:
prediction_50_user_pd_df = prediction_50.toPandas()[['userId','rating','prediction']].set_index('userId')
prediction_50_user_pd_df['prediction'] = prediction_50_user_pd_df['prediction'].apply(lambda x: round_to_5scale(x))
display(prediction_50_user_pd_df.plot(style=['o','rx']))

### Apply model to all data and see the performance

In [66]:
#Best_model RMSE
alldata=best_model.transform(movie_ratings_df)
rmse = evaluator.evaluate(alldata)
print ("RMSE = "+str(rmse))

RMSE = 0.6975433153657589

In [67]:
#Best_model_50 RMSE
alldata=best_model_50.transform(movie_ratings_df)
rmse = evaluator.evaluate(alldata)
print ("RMSE = "+str(rmse))

RMSE = 0.6874073412892818

In [68]:
alldata.registerTempTable("alldata")

In [69]:
%sql 
SELECT *
FROM movies
	JOIN alldata ON movies.movieId = alldata.movieId

movieId title genres userId movieId rating prediction 148 Awfully Big Adventure, An (1995) Drama 191 148 5.0 4.855351 471 Hudsucker Proxy, The (1994) Comedy 133 471 4.0 2.416814 471 Hudsucker Proxy, The (1994) Comedy 597 471 2.0 3.3838356 471 Hudsucker Proxy, The (1994) Comedy 385 471 4.0 2.9433992 471 Hudsucker Proxy, The (1994) Comedy 436 471 3.0 3.3313468 471 Hudsucker Proxy, The (1994) Comedy 602 471 4.0 3.1600466 471 Hudsucker Proxy, The (1994) Comedy 91 471 1.0 2.5822668 471 Hudsucker Proxy, The (1994) Comedy 409 471 3.0 3.0681155 471 Hudsucker Proxy, The (1994) Comedy 372 471 3.0 2.7291358 471 Hudsucker Proxy, The (1994) Comedy 599 471 2.5 2.4970746 471 Hudsucker Proxy, The (1994) Comedy 603 471 4.0 3.0207326 471 Hudsucker Proxy, The (1994) Comedy 182 471 4.5 3.293565 471 Hudsucker Proxy, The (1994) Comedy 218 471 4.0 3.2036498 471 Hudsucker Proxy, The (1994) Comedy 474 471 3.0 3.162521 471 Hudsucker Proxy, The (1994) Comedy 500 471 1.0 2.292801 471 Hudsucker Proxy, The (1994) Comedy 57 471 3.0 3.1477387 471 Hudsucker Proxy, The (1994) Comedy 462 471 2.5 2.4053903 471 Hudsucker Proxy, The (1994) Comedy 387 471 3.0 2.6148915 471 Hudsucker Proxy, The (1994) Comedy 610 471 4.0 3.330954 471 Hudsucker Proxy, The (1994) Comedy 217 471 2.0 2.7944257 471 Hudsucker Proxy, The (1994) Comedy 555 471 3.0 3.4999354 471 Hudsucker Proxy, The (1994) Comedy 176 471 5.0 3.5753264 471 Hudsucker Proxy, The (1994) Comedy 520 471 5.0 3.5087552 471 Hudsucker Proxy, The (1994) Comedy 136 471 4.0 3.414276 471 Hudsucker Proxy, The (1994) Comedy 171 471 3.0 3.9610012 471 Hudsucker Proxy, The (1994) Comedy 273 471 5.0 3.99234 471 Hudsucker Proxy, The (1994) Comedy 448 471 4.0 3.0048363 471 Hudsucker Proxy, The (1994) Comedy 312 471 4.0 3.140611 471 Hudsucker Proxy, The (1994) Comedy 216 471 3.0 2.9039948 471 Hudsucker Proxy, The (1994) Comedy 411 471 4.0 3.1374419 471 Hudsucker Proxy, The (1994) Comedy 287 471 4.5 2.2807083 471 Hudsucker Proxy, The (1994) Comedy 32 471 3.0 3.6704168 471 Hudsucker Proxy, The (1994) Comedy 469 471 5.0 2.996797 471 Hudsucker Proxy, The (1994) Comedy 414 471 5.0 3.1374927 471 Hudsucker Proxy, The (1994) Comedy 608 471 1.5 2.3591225 471 Hudsucker Proxy, The (1994) Comedy 426 471 5.0 4.0650706 471 Hudsucker Proxy, The (1994) Comedy 541 471 3.0 3.3931797 471 Hudsucker Proxy, The (1994) Comedy 260 471 4.5 3.1112723 471 Hudsucker Proxy, The (1994) Comedy 373 471 5.0 2.8377407 471 Hudsucker Proxy, The (1994) Comedy 357 471 3.5 3.6460202 471 Hudsucker Proxy, The (1994) Comedy 104 471 4.5 3.7129288 496 What Happened Was... (1994) Comedy|Drama|Romance|Thriller 191 496 5.0 4.855351 833 High School High (1996) Comedy 44 833 2.0 2.0408332 833 High School High (1996) Comedy 599 833 1.5 1.4265251 833 High School High (1996) Comedy 609 833 3.0 2.0049617 833 High School High (1996) Comedy 307 833 1.0 1.1141282 833 High School High (1996) Comedy 492 833 4.0 2.374479 833 High School High (1996) Comedy 608 833 0.5 0.9945636 1088 Dirty Dancing (1987) Drama|Musical|Romance 463 1088 3.5 3.1402903 1088 Dirty Dancing (1987) Drama|Musical|Romance 159 1088 4.0 3.3042424 1088 Dirty Dancing (1987) Drama|Musical|Romance 606 1088 3.0 3.5096939 1088 Dirty Dancing (1987) Drama|Musical|Romance 599 1088 2.5 2.61794 1088 Dirty Dancing (1987) Drama|Musical|Romance 111 1088 3.0 3.28575 1088 Dirty Dancing (1987) Drama|Musical|Romance 47 1088 4.0 2.9853864 1088 Dirty Dancing (1987) Drama|Musical|Romance 177 1088 3.5 3.6191282 1088 Dirty Dancing (1987) Drama|Musical|Romance 132 1088 4.0 2.9138288 1088 Dirty Dancing (1987) Drama|Musical|Romance 474 1088 3.5 3.274083 1088 Dirty Dancing (1987) Drama|Musical|Romance 20 1088 4.5 3.6362047 1088 Dirty Dancing (1987) Drama|Musical|Romance 169 1088 4.5 4.152839 1088 Dirty Dancing (1987) Drama|Musical|Romance 479 1088 4.0 3.3542252 1088 Dirty Dancing (1987) Drama|Musical|Romance 64 1088 4.0 3.3499837 1088 Dirty Dancing (1987) Drama|Musical|Romance 554 1088 5.0 4.1497188 1088 Dirty Dancing (1987) Drama|Musical|Romance

In [70]:
alldata_pd_df = alldata.toPandas()[['movieId','rating','prediction']].set_index('movieId')
alldata_pd_df.plot(style=['o','rx'])
display()

In [71]:
alldata_pd_df['prediction'] = alldata_pd_df['prediction'].apply(lambda x: round_to_5scale(x))
alldata_pd_df.plot(style=['o','rx'])
display()

In [72]:
alldata_user_pd_df = alldata.toPandas()[['userId','rating','prediction']].set_index('userId')
alldata_user_pd_df['prediction'] = alldata_user_pd_df['prediction'].apply(lambda x: round_to_5scale(x))
display(alldata_user_pd_df.plot(style=['o','rx']))

### Let's look more closely on the user and movie data to decide which user to recommend.

In [74]:
ratings_info_df = movie_ratings_df.groupBy('movieId').avg('rating')
movie_ratings_count= movie_ratings_df.groupBy('movieId').count()
ratings_info_df = ratings_info_df.join(movie_ratings_count,'movieId','left').join(movies_df,'movieId','left')
ratings_info_df.show(3)

+-------+-----------------+-----+--------------------+--------------------+
movieId| avg(rating)|count| title| genres|
+-------+-----------------+-----+--------------------+--------------------+
 1580|3.487878787878788| 165|Men in Black (a.k...|Action|Comedy|Sci-Fi|
 2366| 3.64| 25| King Kong (1933)|Action|Adventure|...|
 3175| 3.58| 75| Galaxy Quest (1999)|Adventure|Comedy|...|
+-------+-----------------+-----+--------------------+--------------------+
only showing top 3 rows

### UserId order by rating count

In [76]:
ratings_df.groupBy("userId").count().toPandas().sort_values(by = 'count',ascending=False).head()

userId 
 count 
 
 
 
 
 345 
 414 
 2698 
 
 
 472 
 599 
 2478 
 
 
 280 
 474 
 2108 
 
 
 28 
 448 
 1864 
 
 
 193 
 274 
 1346

We will recommend user 414 and 599 as they have the highest rating count. This means the recommendations to them would be more reliable.

### Histogram of rating counts

In [79]:
display(ratings_info_df.toPandas()['count'].hist(bins=50,log= True))

### Histogram of the ratings

In [81]:
display(ratings_info_df.toPandas()['avg(rating)'].hist(bins=50))

In [82]:
fig, ax = plt.subplots()
g = sns.jointplot(x='avg(rating)', y='count', data=ratings_info_df.toPandas())
display(g.fig)

## Recommend moive to users with id: 414, 599.

In [84]:
# use the recommendation function of ALS
ALS_recommendations = best_model.recommendForAllUsers(10)
ALS_recommendations.filter(ALS_recommendations['userId'] == 599).show()

+------+--------------------+
userId| recommendations|
+------+--------------------+
 599|[[96004, 4.252195...|
+------+--------------------+

We need to process the above dataframe for readability.

In [86]:
# Data procesing of the ALS_recommendations dataframe
from pyspark.sql.functions import explode,col
recommendations_df = (ALS_recommendations\
                      .select("userId",\
                              explode("recommendations")\
                              .alias("recommendation"))\
                      .select("userId", "recommendation.movieId",\
                              col("recommendation.rating")\
                              .alias('prediction')))
recommendations_df.show(3)

+------+-------+----------+
userId|movieId|prediction|
+------+-------+----------+
 471| 6818| 5.1408505|
 471| 8477| 5.0905204|
 471| 3379| 4.981934|
+------+-------+----------+
only showing top 3 rows

In [87]:
# we only recommend movies that have not been watched by users before
recommendations_df = recommendations_df.join(movies_df,["movieId"],"left").join(ratings_df,['movieId','UserId'],'left')
recommendations_df = recommendations_df.drop('timestamp')
recommendations_df = recommendations_df.filter(ratings_df.rating.isNull())
recommendations_df.show()

+-------+------+----------+--------------------+--------------------+------+
movieId|userId|prediction| title| genres|rating|
+-------+------+----------+--------------------+--------------------+------+
 6818| 471| 5.1408505|Come and See (Idi...| Drama|War| null|
 8477| 471| 5.0905204| Jetée, La (1962)| Romance|Sci-Fi| null|
 3379| 471| 4.981934| On the Beach (1959)| Drama| null|
 96004| 471| 4.981934|Dragon Ball Z: Th...|Action|Adventure|...| null|
 5490| 471| 4.858012| The Big Bus (1976)| Action|Comedy| null|
 5915| 471| 4.858012|Victory (a.k.a. E...| Action|Drama|War| null|
 132333| 471| 4.858012| Seve (2014)| Documentary|Drama| null|
 40491| 471| 4.846524|Match Factory Gir...| Comedy|Drama| null|
 26326| 471| 4.6848273|Holy Mountain, Th...| Drama| null|
 102217| 471| 4.677115|Bill Hicks: Revel...| Comedy| null|
 96004| 463| 5.278501|Dragon Ball Z: Th...|Action|Adventure|...| null|
 3379| 463| 5.278501| On the Beach (1959)| Drama| null|
 60943| 463| 5.0922623| Frozen River (2008)| Drama| null|
 3567| 463| 4.9940257| Bossa Nova (2000)|Comedy|Drama|Romance| null|
 33649| 463| 4.925018| Saving Face (2004)|Comedy|Drama|Romance| null|
 92494| 463| 4.8854833|Dylan Moran: Mons...| Comedy|Documentary| null|
 102217| 463| 4.8854833|Bill Hicks: Revel...| Comedy| null|
 33779| 463| 4.8854833|Eddie Izzard: Dre...| Comedy| null|
 171495| 463| 4.8617916| Cosmos| (no genres listed)| null|
 93988| 463| 4.814151|North & South (2004)| Drama|Romance| null|
+-------+------+----------+--------------------+--------------------+------+
only showing top 20 rows

We will only recommend movies that the users haven't watched/rated.

movies recommendations for user 414

In [89]:
recommendations_df.filter(recommendations_df['userId'] == 414).show()

+-------+------+----------+--------------------+--------------------+------+
movieId|userId|prediction| title| genres|rating|
+-------+------+----------+--------------------+--------------------+------+
 96004| 414| 5.301862|Dragon Ball Z: Th...|Action|Adventure|...| null|
 3379| 414| 5.301862| On the Beach (1959)| Drama| null|
 6818| 414| 5.1921563|Come and See (Idi...| Drama|War| null|
 8477| 414| 5.1837234| Jetée, La (1962)| Romance|Sci-Fi| null|
 132333| 414| 5.026645| Seve (2014)| Documentary|Drama| null|
 5915| 414| 5.026645|Victory (a.k.a. E...| Action|Drama|War| null|
 5490| 414| 5.026645| The Big Bus (1976)| Action|Comedy| null|
 4495| 414| 4.941838|Crossing Delancey...| Comedy|Romance| null|
 92494| 414| 4.9203644|Dylan Moran: Mons...| Comedy|Documentary| null|
 102217| 414| 4.9203644|Bill Hicks: Revel...| Comedy| null|
+-------+------+----------+--------------------+--------------------+------+

movies recommendations for user 599

In [91]:
recommendations_df.filter(recommendations_df['userId'] == 599).show()

+-------+------+----------+--------------------+------------------+------+
movieId|userId|prediction| title| genres|rating|
+-------+------+----------+--------------------+------------------+------+
 5490| 599| 3.9496071| The Big Bus (1976)| Action|Comedy| null|
 5915| 599| 3.9496071|Victory (a.k.a. E...| Action|Drama|War| null|
 132333| 599| 3.9496071| Seve (2014)| Documentary|Drama| null|
 60943| 599| 3.9102483| Frozen River (2008)| Drama| null|
 4495| 599| 3.8764865|Crossing Delancey...| Comedy|Romance| null|
 6818| 599| 3.8465486|Come and See (Idi...| Drama|War| null|
 92494| 599| 3.8462806|Dylan Moran: Mons...|Comedy|Documentary| null|
+-------+------+----------+--------------------+------------------+------+

In [92]:
# Another way is to recommendForUserSubset function
#users = ALS_recommendations.filter(ALS_recommendations['userId'] == 575)
#ALS_recommendations_target = best_model.recommendForUserSubset(users,1)
#ALS_recommendations_target.show()

## Find the similar moives for moive with id: 464, 471

The similarites of different movies can be recognized from correlations of their user ratings. Let's say user A and B both gave movie 1,2 5 star rating. This indicates that the movie 1 and 2 might be highily similar.

However, we have a challenge in that some of the movies have very few ratings and may end up having high correlation simply because one or two people gave them a 5 star rating. We can fix this by setting a threshold for the number of ratings. From the histogram earlier we saw a sharp decline in number of ratings from 100. Therefore we will choose this as our threshold.

In [95]:
movie_matrix = movie_ratings_df.toPandas().pivot_table(index='userId', columns='movieId', values='rating')

/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "

In [96]:
def find_similar_movie(x):
  movie_x_rating = movie_matrix[x]
  similar_to_x=movie_matrix.corrwith(movie_x_rating).reset_index(level=0)
  similar_to_x.dropna(axis = 0,how = 'any',inplace=True)
  similar_to_x.columns = ['movieId','correlation']
  
  similar_to_x_df = sqlContext.createDataFrame(similar_to_x)
  similar_to_x_movie = similar_to_x_df.join(ratings_info_df,'movieId','left').toPandas()[['movieId','correlation','title','count']]
  res = similar_to_x_movie[similar_to_x_movie['count']>100].sort_values(by = 'correlation',ascending = False)
  return similar_to_x,res

We will only find similar movies that have been rated more than 100 times.

movies similar to movie 471

In [98]:
# Movies similar to 471
corr_471,similar_to_471_movie = find_similar_movie(471)

/databricks/python/lib/python3.7/site-packages/numpy/lib/function_base.py:2522: RuntimeWarning: Degrees of freedom <= 0 for slice
 c = cov(x, y, rowvar)
/databricks/python/lib/python3.7/site-packages/numpy/lib/function_base.py:2451: RuntimeWarning: divide by zero encountered in true_divide
 c *= np.true_divide(1, fact)
/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "

In [99]:
corr_471.plot(kind='scatter',x = 'movieId',y='correlation')
display()

In [100]:
similar_to_471_movie.head()

movieId 
 correlation 
 title 
 count 
 
 
 
 
 2426 
 79132 
 0.973124 
 Inception (2010) 
 143 
 
 
 1788 
 48516 
 0.724861 
 Departed, The (2006) 
 107 
 
 
 2576 
 33794 
 0.677430 
 Batman Begins (2005) 
 116 
 
 
 1350 
 1222 
 0.608101 
 Full Metal Jacket (1987) 
 102 
 
 
 1229 
 364 
 0.570549 
 Lion King, The (1994) 
 172

movies similar to movie 464

In [102]:
# Movies similar to 464
corr_464,similar_to_464_movie = find_similar_movie(464)

/databricks/python/lib/python3.7/site-packages/numpy/lib/function_base.py:2522: RuntimeWarning: Degrees of freedom <= 0 for slice
 c = cov(x, y, rowvar)
/databricks/python/lib/python3.7/site-packages/numpy/lib/function_base.py:2451: RuntimeWarning: divide by zero encountered in true_divide
 c *= np.true_divide(1, fact)
/databricks/python/lib/python3.7/site-packages/pyarrow/__init__.py:152: UserWarning: pyarrow.open_stream is deprecated, please use pyarrow.ipc.open_stream
 warnings.warn("pyarrow.open_stream is deprecated, please use "

In [103]:
corr_464.plot(kind='scatter',x = 'movieId',y='correlation')
display()

In [104]:
similar_to_464_movie.head()

movieId 
 correlation 
 title 
 count 
 
 
 
 
 362 
 5816 
 1.000000 
 Harry Potter and the Chamber of Secrets (2002) 
 102 
 
 
 528 
 4896 
 0.904534 
 Harry Potter and the Sorcerer's Stone (a.k.a. ... 
 107 
 
 
 1465 
 58559 
 0.866025 
 Dark Knight, The (2008) 
 149 
 
 
 450 
 253 
 0.853913 
 Interview with the Vampire: The Vampire Chroni... 
 109 
 
 
 389 
 7153 
 0.845154 
 Lord of the Rings: The Return of the King, The... 
 185

## Write the report 
motivation
1. step1
2. step2
3. step3
4. step4  
output and conclusion

## Report
In this project, the movie data from movie lens data set which includes about 600 users and 9500 movies were analyzed to gain insights on movie recomendations to users and finding similar movies. At the beginning, we calculate the sparisity of the movie ratings which is 98.3%, this tells us that the ratings dataframe is mostly empty, which brings significance to predict the user ratings from what we have.

To achieve the goal,the data was analyzed on Spark platform from perfoming data cleaning,processing to model training with Alternating Least Squares (ALS) algorithm.During which, grid search and cross validation were applied to tune the hyperparameters. It is found that using large rank and iterzations would help achieve a low RMSE. Finally, we choose a rank of 10, iterate 50 times and regPram = 0.15,a RMSE of 0.69 was achieved.This means that on average the model predicts 0.69 above or below values of the original ratings matrix.

By successsfully predicting the ratings using the best model, we not only fill the rating dataframe and recommend our users with movies they have never watched, but also find similar movies through their correlations. This brings huge business value to the company.